server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java

```diff
@@ -224,14 +244,23 @@ public final void write(final T state, final Path... locations) throws WriteStat
             copyStateToExtraLocations(directories, tmpFileName);
             performRenames(tmpFileName, fileName, directories);
             performStateDirectoriesFsync(directories);
+        } catch (WriteStateException e) {
+            if (cleanup) {
+                cleanupOldFiles(oldGenerationId, locations);
```

Comment:
Please see my comment above why I've done it in this way. Shortly, the reason why I'm using writeManifestAndCleanup, because I don't know manifest file generation. If manifest write succeeds I get current generation returned by the method call, if it fails - I don't have it.
In general, I don't think there is a problem with other callers calling writeAndCleanup, because they don't use the concept of manifest and they want to keep the previously written latest generation.
If you're really afraid of introducing a regression, I can suggest adding "oldGenerationId" to WriteStateException, that when writing manifest failed we can learn which file to delete. IMHO, this is not needed and cleaning leaving only previous generation of state files for all uses of writeAndCleanup is the right thing to do.